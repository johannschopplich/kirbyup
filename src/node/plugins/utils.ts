export const multilineCommentsRE = /\/\*(.|[\r\n])*?\*\//gm
export const singlelineCommentsRE = /\/\/.*/g

export const HMR_RUNTIME_ID = '\0plugin-vue2:hmr-runtime'
export const JSX_HMR_RUNTIME_ID = 'plugin-vue2-jsx:hmr-runtime'

export const isHmrRuntimeId = (id: string) => id === HMR_RUNTIME_ID || id === JSX_HMR_RUNTIME_ID

/**
 * This code is injected into the HMR runtime of plugin-vue2(-jsx).
 *
 * All .vue components register themselves once with the HMR runtime, so their exported
 * component definitions can be stored in a map, alongside the rendered component instances
 * that are based off this definition. When a module is updated, the runtime applies all
 * changes from the updated module to the stored definition, then re-renders the instances.
 *
 * {
 *   [id]: { options: ComponentDefinition, instances: [...] }
 * }
 *
 * However, in some cases (sections and blocks) Kirby does not actually register the
 * object that is exported from a .vue file (and stored as definition) as component,
 * instead it creates a new object and merges the definition from the .vue file in:
 * https://github.com/getkirby/kirby/blob/main/panel/public/js/plugins.js#L19-L22
 * After changes, the runtime updates the definition and re-renders the instances, but since
 * they are derived from the object created by Kirby, not the stored definition, nothing happens.
 * To fix that, we wrap rerender() and reload() so that before applying the updates, we first check
 * if the updated definition belongs to a component added by a Kirby plugin. To do so, we can check
 * whether the __file (added by plugin-vue2) or __hmrId (added by plugin-vue2-jsx) properties of the
 * updated module and the plugin component match. If so, we look up the component definition that is
 * *actually* used by component instances rendered on the page (window.panel.$vue.$options.components)
 * and if it differs from the one stored in the HMR runtime's map, we updates the map's reference.
 *
 * We also check the component name and add a $_isSection flag if it's k-something-section, because
 * section components are hard to detect and need special treatment in $_applyKirbyModifications.
 *
 * $_applyKirbyModifications:
 *
 * Kirby modifies component definitions before registering components.
 * This includes adding the section mixin to section components,
 * giving templates priority over render functions if both exist
 * and resolving component names in "extends" to their definition object:
 * https://github.com/getkirby/kirby/blob/main/panel/src/config/plugins.js
 * When a module is reloaded, Vue receives a fresh component definition that is
 * missing these modifications. We need to re-apply them, else the runtime will
 * prune them when patching the stored definition to match the newer one.
 *
 * The call to $_applyKirbyModifications() is injected into __VUE_HMR_RUNTIME__.reload()
 * at the appropriate position using a RegExp in the Vite plugin's transform method.
 */

export const __INJECTED_HMR_CODE__ = `
/** - injected by kirbyup - */
for (const methodName of ['rerender', 'reload']) {
  const original = __VUE_HMR_RUNTIME__[methodName]

  __VUE_HMR_RUNTIME__[methodName] = function (id, updatedDef) {
    const key = updatedDef?.__file ? '__file' : updatedDef?.__hmrId ? '__hmrId' : null

    if (key) {
      const pluginComponents = window.panel.plugins.components
      const usedComponentDefs = window.panel.$vue._vnode.componentInstance.$options.components

      for (const componentName in pluginComponents) {
        if (updatedDef[key] === pluginComponents[componentName][key]) {
          const usedDefinition = usedComponentDefs[componentName].options

          if (map[id].options !== usedDefinition)
            map[id].options = usedDefinition
          
          if (typeof map[id].options.$_isSection !== 'boolean')
            map[id].options.$_isSection = /^k-.*-section$/.test(componentName)

          break
        }
      }
    }

    return original.apply(this, arguments)
  }
}

function $_applyKirbyModifications(activeDef, newDef) {
  const usedComponentDefs = window.panel.$vue.$options.components

  if (newDef.template)
    newDef.render = null

  if (activeDef.$_isSection)
    newDef.$_isSection = true
  if (newDef.$_isSection && !newDef.mixins?.[0]?.methods?.load)
    newDef.mixins = [activeDef.mixins[0], ...(newDef.mixins || [])]

  if (typeof newDef.extends === 'string') {
    if (newDef.extends === activeDef.extends?.options?.name) {
      newDef.extends = activeDef.extends
    }
    else if (usedComponentDefs[newDef.extends]) {
      newDef.extends = usedComponentDefs[newDef.extends].extend({
        options: newDef,
        components: { ...usedComponentDefs, ...(newDef.components || {}) },
      })
    }
    else { newDef.extends = null }
  }
}
/** -- */
`
