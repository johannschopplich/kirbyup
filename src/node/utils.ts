import { Buffer } from 'node:buffer'
import { promisify } from 'node:util'
import { gzip } from 'node:zlib'
import { consola } from 'consola'
import { colors } from 'consola/utils'
import { normalize, relative, resolve } from 'pathe'

export function toArray<T>(array?: T | T[]): T[] {
  array ??= []
  return Array.isArray(array) ? array : [array]
}

const compress = promisify(gzip)

export async function getCompressedSize(code: string | Uint8Array) {
  const size = (await compress(typeof code === 'string' ? code : Buffer.from(code))).length / 1024
  return ` / gzip: ${size.toFixed(2)} KiB`
}

export async function printFileInfo(
  {
    root,
    outDir,
    filePath,
    content,
    type,
    maxLength,
  }: {
    root: string
    outDir: string
    filePath: string
    content: string
    type: string
    maxLength: number
  },
) {
  const prettyOutDir = `${normalize(relative(root, resolve(root, outDir)))}/`
  const kibs = content.length / 1024
  const compressedSize = await getCompressedSize(content)
  const writeColor = type === 'chunk' ? colors.cyan : colors.magenta

  consola.log(
    colors.white(colors.dim(prettyOutDir))
    + writeColor(filePath.padEnd(maxLength + 2))
    + colors.dim(`${kibs.toFixed(2)} kB${compressedSize}`),
  )
}
