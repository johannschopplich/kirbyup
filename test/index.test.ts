import { resolve } from 'pathe'
import { remove } from 'fs-extra'
import { afterAll, beforeAll, expect, it } from 'vitest'
import { cacheDir, runCli } from './utils'

beforeAll(async () => {
  await remove(cacheDir)
})

afterAll(async () => {
  await remove(cacheDir)
})

it('builds index.js', async () => {
  const { output } = await runCli({
    'src/input.js': 'import foo from \'./foo\'\nexport default foo',
    'src/foo.js': 'export default \'bar\'',
  })

  expect(output).toMatchSnapshot()
})

it('builds index.css', async () => {
  const { output, getFileContent } = await runCli({
    'src/input.js': 'import \'./input.css\'',
    'src/input.css': '.foo { content: "bar"; }',
  })

  expect(output).toMatchSnapshot()

  const css = await getFileContent('index.css')
  expect(css).toMatchSnapshot()
})

it('supports resolve aliases', async () => {
  const { output } = await runCli({
    'src/input.js': 'import foo from \'~/foo\'\nexport default foo',
    'src/foo.js': 'export default \'bar\'',
  })

  expect(output).toMatchSnapshot()
})

it('supports built-in env variables', async () => {
  const { output } = await runCli({
    'src/input.js': 'export const mode = import.' + 'meta.env.MODE',
  })

  expect(output).toMatchSnapshot()
})

it('supports custom env variables', async () => {
  const { output } = await runCli({
    '.env': 'KIRBYUP_FOO=bar',
    'src/input.js': 'export const foo = import.' + 'meta.env.KIRBYUP_FOO',
  })

  expect(output).toMatchSnapshot()
})

it('supports postcss plugins', async () => {
  const { output, getFileContent } = await runCli({
    'src/input.js': 'import \'./input.css\'',
    'src/input.css': `
      .foo { inset: logical 0 5px 10px; }
      .bar:dir(rtl) { margin-right: 10px; }
    `,
  })

  expect(output).toMatchSnapshot()

  const css = await getFileContent('index.css')
  expect(css).toMatchSnapshot()
})

it('builds panel plugins', async () => {
  const { output } = await runCli({
    'src/input.js': `
      import Demo from './fields/demo.js'
      window.panel.plugin('kirbyup/test', {
        fields: {
          demo: Demo
        }
      })
    `,
    'src/fields/demo.js': 'export default { extends: \'k-info-field\' }',
  })

  expect(output).toMatchSnapshot()
})

it('compiles vue templates', async () => {
  const { output } = await runCli({
    'src/input.js': `
      import DemoSection from './fields/DemoSection.vue'
      window.panel.plugin('kirbyup/test', {
        sections: {
          demo: DemoSection
        }
      })
    `,
    'src/fields/DemoSection.vue': `
      <template>
        <section class="k-demo-section">
          <header class="k-section-header">
            <h2 class="k-headline">Custom section</h2>
          </header>
        </section>
      </template>

      <script>
      export default {
        created() {
          console.log('created')
        }
      };
      </script>

      <style>
      .k-demo-section {
        padding: 20px;
      }
      </style>
    `,
  })

  expect(output).toMatchSnapshot()
})

it('supports auto-importing components', async () => {
  const { output } = await runCli({
    'src/input.js': `
      import { kirbyup } from '${resolve(__dirname, '../dist/plugin.mjs')}'

      window.panel.plugin('kirbyup/example', {
        blocks: kirbyup.import('./components/blocks/*.vue')
      })
    `,
    'src/components/blocks/Foo.vue': '<template><k-header>Foo</k-header></template>',
    'src/components/blocks/Bar.vue': '<template><k-header>Bar</k-header></template>',
  })

  expect(output).toMatchSnapshot()
})

it('supports kirbyup.config.js with object', async () => {
  const { output } = await runCli({
    'src/input.js': 'import foo from \'__ALIAS__/foo\'\nexport default foo',
    'src/foo.js': 'export default \'bar\'',
    'kirbyup.config.js': `
      import { resolve } from 'path'
      export default {
        alias: {
          '__ALIAS__/': resolve(__dirname, 'src') + '/'
        },
        extendViteConfig: {
          build: {
            lib: {
              name: 'test'
            }
          }
        }
      }
    `,
  })

  expect(output).toMatchSnapshot()
})

it('supports kirbyup.config.js with function', async () => {
  const { output } = await runCli({
    'src/input.js': 'import foo from \'__ALIAS__/foo\'\nexport default foo',
    'src/foo.js': 'export default \'bar\'',
    'kirbyup.config.js': `
      import { resolve } from 'path'
      import { defineConfig } from '${resolve(__dirname, '../dist/config.mjs')}'
      export default defineConfig({
        alias: {
          '__ALIAS__/': resolve(__dirname, 'src') + '/'
        },
        extendViteConfig: {
          build: {
            lib: {
              name: 'test'
            }
          }
        }
      })
    `,
  })

  expect(output).toMatchSnapshot()
})
