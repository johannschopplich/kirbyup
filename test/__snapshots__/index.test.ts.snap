// Vitest Snapshot v1

exports[`builds index.css 1`] = `
"(function(){\\"use strict\\";const t=\\"\\"})();
"
`;

exports[`builds index.css 2`] = `
".foo{content:\\"bar\\"}
"
`;

exports[`builds index.js 1`] = `
"var kirbyupExport=function(){\\"use strict\\";return\\"bar\\"}();
"
`;

exports[`builds panel plugins 1`] = `
"(function(){\\"use strict\\";const e={extends:\\"k-info-field\\"};window.panel.plugin(\\"kirbyup/test\\",{fields:{demo:e}})})();
"
`;

exports[`compiles vue templates 1`] = `
"(function(){\\"use strict\\";const g=\\"\\";function l(o,n,p,m,i,r,a,C){var e=typeof o==\\"function\\"?o.options:o;n&&(e.render=n,e.staticRenderFns=p,e._compiled=!0),m&&(e.functional=!0),r&&(e._scopeId=\\"data-v-\\"+r);var s;if(a?(s=function(t){t=t||this.$vnode&&this.$vnode.ssrContext||this.parent&&this.parent.$vnode&&this.parent.$vnode.ssrContext,!t&&typeof __VUE_SSR_CONTEXT__<\\"u\\"&&(t=__VUE_SSR_CONTEXT__),i&&i.call(this,t),t&&t._registeredComponents&&t._registeredComponents.add(a)},e._ssrRegister=s):i&&(s=C?function(){i.call(this,(e.functional?this.parent:this).$root.$options.shadowRoot)}:i),s)if(e.functional){e._injectStyles=s;var R=e.render;e.render=function($,f){return s.call(f),R($,f)}}else{var _=e.beforeCreate;e.beforeCreate=_?[].concat(_,s):[s]}return{exports:o,options:e}}const c={created(){console.log(\\"created\\")}};var d=function(){var n=this;return n._self._c,n._m(0)},u=[function(){var o=this,n=o._self._c;return n(\\"section\\",{staticClass:\\"k-demo-section\\"},[n(\\"header\\",{staticClass:\\"k-section-header\\"},[n(\\"h2\\",{staticClass:\\"k-headline\\"},[o._v(\\"Custom section\\")])])])}],v=l(c,d,u,!1,null,null,null,null);const h=v.exports;window.panel.plugin(\\"kirbyup/test\\",{sections:{demo:h}})})();
"
`;

exports[`supports auto-importing components 1`] = `
"(function(){\\"use strict\\";function a(n,o,_,i,s,l,f,S){var e=typeof n==\\"function\\"?n.options:n;o&&(e.render=o,e.staticRenderFns=_,e._compiled=!0),i&&(e.functional=!0),l&&(e._scopeId=\\"data-v-\\"+l);var t;if(f?(t=function(r){r=r||this.$vnode&&this.$vnode.ssrContext||this.parent&&this.parent.$vnode&&this.parent.$vnode.ssrContext,!r&&typeof __VUE_SSR_CONTEXT__<\\"u\\"&&(r=__VUE_SSR_CONTEXT__),s&&s.call(this,r),r&&r._registeredComponents&&r._registeredComponents.add(f)},e._ssrRegister=t):s&&(t=S?function(){s.call(this,(e.functional?this.parent:this).$root.$options.shadowRoot)}:s),t)if(e.functional){e._injectStyles=t;var T=e.render;e.render=function(j,c){return t.call(c),T(j,c)}}else{var u=e.beforeCreate;e.beforeCreate=u?[].concat(u,t):[t]}return{exports:n,options:e}}const d={};var v=function(){var o=this,_=o._self._c;return _(\\"k-header\\",[o._v(\\"Bar\\")])},p=[],b=a(d,v,p,!1,null,null,null,null);const m=b.exports,h=Object.freeze(Object.defineProperty({__proto__:null,default:m},Symbol.toStringTag,{value:\\"Module\\"})),g={};var C=function(){var o=this,_=o._self._c;return _(\\"k-header\\",[o._v(\\"Foo\\")])},O=[],$=a(g,C,O,!1,null,null,null,null);const k=$.exports,R=Object.freeze(Object.defineProperty({__proto__:null,default:k},Symbol.toStringTag,{value:\\"Module\\"})),y=n=>n.substring(n.lastIndexOf(\\"/\\")+1,n.lastIndexOf(\\".\\")).toLowerCase(),F=Object.freeze({import(n){return Object.entries(n).reduce((o,[_,i])=>(o[y(_)]=i.default,o),{})}});window.panel.plugin(\\"kirbyup/example\\",{blocks:F.import(Object.assign({\\"./components/blocks/Bar.vue\\":h,\\"./components/blocks/Foo.vue\\":R}))})})();
"
`;

exports[`supports built-in env variables 1`] = `
"var kirbyupExport=function(e){\\"use strict\\";const o=\\"production\\";return e.mode=o,Object.defineProperties(e,{__esModule:{value:!0},[Symbol.toStringTag]:{value:\\"Module\\"}}),e}({});
"
`;

exports[`supports custom env variables 1`] = `
"var kirbyupExport=function(e){\\"use strict\\";const o={BASE_URL:\\"/\\",MODE:\\"production\\",DEV:!1,PROD:!0}.KIRBYUP_FOO;return e.foo=o,Object.defineProperties(e,{__esModule:{value:!0},[Symbol.toStringTag]:{value:\\"Module\\"}}),e}({});
"
`;

exports[`supports kirbyup.config.js with function 1`] = `
"var test=function(){\\"use strict\\";return\\"bar\\"}();
"
`;

exports[`supports kirbyup.config.js with object 1`] = `
"var test=function(){\\"use strict\\";return\\"bar\\"}();
"
`;

exports[`supports postcss plugins 1`] = `
"(function(){\\"use strict\\";const t=\\"\\"})();
"
`;

exports[`supports postcss plugins 2`] = `
".foo{top:0;left:5px;bottom:10px;right:5px}[dir=rtl] .bar{margin-right:10px}
"
`;

exports[`supports resolve aliases 1`] = `
"var kirbyupExport=function(){\\"use strict\\";return\\"bar\\"}();
"
`;
